bluesky/bundlers.py

```diff
@@ -555,63 +553,70 @@ async def collect(self, msg):
 
         Expect message object is
 
-            Msg('collect', obj)
+            Msg('collect', collect_obj)
             Msg('collect', flyer_object, stream=True, return_payload=False)
         """
-        obj = msg.obj
+        collect_obj = msg.obj
 
         if not self.run_is_open:
             # sanity check -- 'kickoff' should catch this and make this
             # code path impossible
             raise IllegalMessageSequence(
                 "A 'collect' message was sent but no " "run is open."
             )
-        self._uncollected.discard(obj)
+        self._uncollected.discard(collect_obj)
 
-        if hasattr(obj, "collect_asset_docs"):
+        if hasattr(collect_obj, "collect_asset_docs"):
             # Resource and Datum documents
-            for name, doc in obj.collect_asset_docs():
+            for name, doc in collect_obj.collect_asset_docs():
                 # Add a 'run_start' field to the resource document on its way out.
                 if name == "resource":
                     doc["run_start"] = self._run_start_uid
                 await self.emit(DocumentNames(name), doc)
 
-        named_data_keys = obj.describe_collect()
-        # e.g., {name_for_desc1: data_keys_for_desc1,
-        #        name for_desc2: data_keys_for_desc2, ...}
+        collect_obj_configuration = {}
+        if hasattr(collect_obj, "read_configuration"):
+            doc_logger.debug("reading configuration from %s", collect_obj)
+            collect_obj_configuration = collect_obj.read_configuration()
+        else:
+            doc_logger.debug("%s has no read_configuration method", collect_obj)
+
         bulk_data = {}
-        local_descriptors = {}  # hashed on obj_read, not (name, objs_read)
-        for stream_name, data_keys in named_data_keys.items():
-            desc_key = stream_name
-            d_objs = frozenset(data_keys)
-            if desc_key not in self._descriptors:
+        local_descriptors = {}  # hashed on objs_read, not (name, objs_read)
+        # collect_obj.describe_collect() returns a dictionary like this:
+        #     {name_for_desc1: data_keys_for_desc1,
+        #      name_for_desc2: data_keys_for_desc2, ...}
+        for stream_name, stream_data_keys in collect_obj.describe_collect().items():
+            d_objs = frozenset(stream_data_keys)
+            if stream_name not in self._descriptors:
                 objs_read = d_objs
```

Comment:
I think this is the source of the confusion. Elsewhere (e.g. in the `save` method) `objs_read` is a `frozenset` of live objects (e.g. Devices). This is a `frozenset` of strings. We should use a different name for it.

Reply:
Is the `frozenset` necessary? Is there a reason not to use `stream_data_keys`?

Reply:
Great question. We make a `frozenset` because are around to use it as a key in a dict. That means it had to be hashable, and mutable objects are not hashable—hence, make a copy in an immutable data structure and proceed with that. Does that answer your question?

Reply:
Thanks. I tend to use a tuple for no good reason in these cases.

Reply:
Yeah, I would normally reach for a tuple first as well, just because it's a more familiar data structure. But we want to make sure that `{'a', 'b'}` == `{'b', 'a'}`, which of course isn't the case for a tuple. We could create a tuple and sort the elements first, but that's easy to mess up if you forget to sort somewhere.